clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -551,18 +538,37 @@ private void reenqueueBatch(ProducerBatch batch, long currentTimeMs) {
     }
 
     private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response) {
+        if (transactionManager != null && transactionManager.hasProducerIdAndEpoch(batch.producerId(), batch.producerEpoch())) {
+            transactionManager.incrementSequenceNumber(batch.topicPartition, batch.recordCount);
+            log.debug("Incremented sequence number for topic-partition {} to {}", batch.topicPartition,
+                    transactionManager.sequenceNumber(batch.topicPartition));
+        }
+
         batch.done(response.baseOffset, response.logAppendTime, null);
         this.accumulator.deallocate(batch);
     }
 
     private void failBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response, RuntimeException exception) {
-        if (transactionManager != null && !transactionManager.isTransactional()
-                && batch.producerId() == transactionManager.producerIdAndEpoch().producerId) {
-            // Reset the transaction state since we have hit an irrecoverable exception and cannot make any guarantees
-            // about the previously committed message. Note that this will discard the producer id and sequence
-            // numbers for all existing partitions.
-            transactionManager.resetProducerId();
+        if (transactionManager != null) {
+            if (exception instanceof OutOfOrderSequenceException && transactionManager.hasProducerId(batch.producerId())) {
+                log.error("The broker received an out of order sequence number for topic-partition " +
+                                "{} at offset {}. This indicates data loss on the broker, and should be investigated.",
+                        batch.topicPartition, response.baseOffset);
+
+                if (transactionManager.isTransactional())
+                    transactionManager.transitionToAbortableError(exception);
+                else
+                    // Reset the transaction state since we have hit an irrecoverable exception and cannot make any guarantees
+                    // about the previously committed message. Note that this will discard the producer id and sequence
+                    // numbers for all existing partitions.
+                    transactionManager.resetProducerId();
+            } else if (transactionManager.isTransactional() || exception instanceof ClusterAuthorizationException) {
```

Comment:
Mentioned offline, but this check would make too many exceptions `fatal`. Instead, we should explicitly check for `INVALID_PRODUCER_EPOCH`, `CLUSTER_AUTHORIZATION_FAILED`, and `TRANSACTIONAL_ID_AUTHORIZATION_FAILED`, and make only those fatal. The rest should be abortable.